### ISSUE TITLE
Support for monitored properties

### ISSUE THREAD

**Issue by dgl-cw** [stlehmann/pyads#261]
Hi

I was using `read_by_name` before and wanted to switch to `get_symbol(...).read()` because of the nicer syntax and not having to pass around the plc object. So I discovered, that the latter failed for monitored properties.

With monitored properties I mean properties that have the [monitoring pragma](https://infosys.beckhoff.com/content/1033/tc3_plc_intro/2529692299.html?id=6668617382077000998).

I couldn't find any documentation on accessing monitored properties over ADS, neither here nor on TwinCAT, so I experimented a bit.

Windows 10
TwinCAT 3.4024.10

With `read_by_name`
| `monitoring` | GVL | Program | Function Block |
| ------------- | ---- | --------- | --------------- |
| `variable` | (✔️) | (✔️) | ❌1808 |
| `call` | ❌1808 | ❌1808 | ✔️ |

❌ is a `ADSError: symbol not found (1808)`
Same for `write_by_name` however with `{attribute 'monitoring' := 'variable'}` writing has not effect.

With `get_symbol(...).read()` or `.write(...)`
| `monitoring` | GVL | Program | Function Block |
| ------------- | ---- | --------- | --------------- |
| `variable` | (✔️) | (✔️) | ❌1808 |
| `call` | ❌1808 | ❌1808  | ❌**1806** |

Mostly the same as above but function block with `call` fails with `ADSError: objects do not match (1806)`
Again `.write(...)` has no effect.

I was mostly interested in the `1806` case so I looked deeper and found that for the affected case `adsGetSymbolInfo` returns an index_group of `0xF019` and an index_offset of `0x0`.
I could not find any documentation on the index group `0xF019`.
As a workaround I tried to get a handle with `plc.get_handle(...)` for the case that index_group was `0xF019` and use that instead. That worked.

I did not look deeper into the case of the `1808` error.

## In summary

I think

* `get_symbol` should not fail, where `read_by_name` succeeds. Thus a workaround should be implemented.
* it would be worth documenting that monitored properties can be read and written through ADS, with the limitations listed above.

**Comment by RobertoRoos** [comment]
Woof, this is still relevant. Just ran into this issue. I am also getting those `objects do not match (1806)` errors.

I would like to add that properties marked with
```
{attribute 'monitoring' := 'call'}
```

...show up in the TwinCAT target browser with the same `index_group` and `index_offset` as the parent function-block instance.

And indeed when I inspect the indices of the `AdsSymbol` instance I see `index_group=0xF019` and `index_offset=0`.

(This is with `TwinCAT v3.1.4026.20`)

EDIT: I am curious what happens on Linux, with `adslib`, if the same issues arise. Unfortunately I don't have easy access to a proper Linux platform now.

**Comment by RobertoRoos** [comment]
I've contacted Beckhoff support about this, asking for the meaning of this group index.
If they confirm that this is a predictable output, we could write a catch for this case into `AdsSymbol`.

EDIT: I can also completely reproduce this issue within TwinCAT:

```
PROGRAM MAIN
VAR
	val				: UDINT;
	
	otherModule		: FB_OtherModule;
	
	fbAdsCall		: ADSRDWRTEX;
	nReadOption		: USINT := 0;
	sWriteBuffer	: STRING;
	sReadBuffer		: ARRAY [1..8] OF DWORD;
END_VAR

----------------------

otherModule();

CASE nReadOption OF
	0:
		fbAdsCall(WRTRD := FALSE);
	
	1:  //  Indices of a regular variable:
		sWriteBuffer := 'MAIN.val';
		
	2:  //  Indices of an FB variable:
		sWriteBuffer := 'MAIN.otherModule.ps_nMyVar';
		
	3:  //  Indices of a monitored property:
		sWriteBuffer := 'MAIN.otherModule.p_nMyVar';

END_CASE

fbAdsCall(
	NETID := '',
	PORT := 851,
	IDXGRP := 16#F009,
	IDXOFFS := 16#0,
	WRITELEN := TO_UDINT(LEN(sWriteBuffer)),
	READLEN := 6 * 4 + 3 * 2, // 6 x 32bit integers + 3 x 16 bit integers (skipping the final buffer)
	SRCADDR := ADR(sWriteBuffer),
	DESTADDR := ADR(sReadBuffer),
	WRTRD := (nReadOption > 0),
);

IF NOT fbAdsCall.BUSY THEN
	nReadOption := 0;
END_IF

// sReadBuffer[2] is the index-group and sReadBuffer[3] is the index-offset
```

**Comment by RobertoRoos** [comment]
I heard back from Beckhoff, they at least confirmed the result. Because of the dynamic nature of the monitored properties, we shouldn't expect to be able to read from static indices at all. So this isn't something that will be easily fixed.

My question is still whether a check on `0xF019` would be robust. But considering this has been constant for the last 5 year, I doubt it would change.

My suggestion is to do the following:
- Add a `warning.warn()` for when the group index has this magic value. Here I prefer a warning over an exception because this is all kind of unofficial and for all we know there is a user somewhere where the magic index is legit.
  - I do want to show _something_, because otherwise the issue is very unclear by itself.
- We document this stuff about monitored vars.
- We add a new argument to the `AdsSymbol` constructor, something like `use_handle`.
  - When `True`, `read/write_by_handle` is used over an index read/write. The handle will we tracked by the symbol object and released at the end.
  - This way a user will have to choose to rely on the handle workaround, which might be best.
  - I dislike adding _even more_ arguments to Symbol, because the parameters are already a bit of a mess. (I regret how I put it together at the time.)

@stlehmann (maybe @chrisbeardy ), would we interested in merging this? If so, I'll throw it together.

**Comment by stlehmann** [comment]
@RobertoRoos seems to me like a reasonable approach. I'm a bit reluctant with modifying the AdsSymbol code because of #298. I still think moving the ADS Symbol approach in an extra repository would be a good way to keep both ways of accessing PLC variables in their own space. (pyads for the basics and pyads-symbols for the more object-oriented approach).

However if you would like to fix this error feel free to create a PR for it. It will be useful one or the other way. 👍

**Comment by RobertoRoos** [comment]
That's a fair point. I'll think a little bit more about it.

**Comment by RobertoRoos** [comment]
In the meantime I've had some back-and-forth with Beckhoff support. I talked through ticket BTS-333875, for future reference.
They acknowledge the 'phenomenon' (whether a bug or not) and though the magic index `0xF019` is not clearly documented, we should assume it at least won't ever change. So I guess we could continue with the plan.
